source/slang/slang-ir-loop-inversion.cpp

```diff
@@ -0,0 +1,305 @@
+#include "slang-ir-loop-inversion.h"
+
+#include "slang-ir-clone.h"
+#include "slang-ir-dominators.h"
+#include "slang-ir-insts.h"
+#include "slang-ir-lower-witness-lookup.h"
+#include "slang-ir-reachability.h"
+#include "slang-ir-ssa-simplification.h"
+#include "slang-ir-util.h"
+#include "slang-ir.h"
+
+namespace Slang
+{
+
+static bool isSameBlockOrTrivialBranch(IRBlock* target, IRBlock* scrutinee)
+{
+    if(target == scrutinee)
+        return true;
+    const auto br = as<IRUnconditionalBranch>(scrutinee->getFirstOrdinaryInst());
+    return br && br->getTargetBlock() == target && br->getArgCount() == 0 && !scrutinee->hasMoreThanOneUse();
+};
+
+static bool isSmallBlock(IRBlock* c)
+{
+    // Somewhat arbitrarily, 4 instructions, enough for:
+    // - Arith
+    // - Comparison
+    // - Negation
+    // - Terminator
+    Int n = 0;
+    for([[maybe_unused]] const auto i : c->getOrdinaryInsts())
+        if(++n > 4)
+            return false;
+    return true;
+}
+
+// Loops are suitable for inversion if:
+// - The loop jumps to a conditional branch which has the break block as one of
+//   its successors (or a trivial break block which we erase) and the other
+//   successor is empty
+// - The conditional block is "small", because we will be duplicating it
+static bool isSuitableForInversion(IRLoop* loop)
+{
+    const auto nextBlock = loop->getTargetBlock();
+    const auto breakBlock = loop->getBreakBlock();
+
+    // The first thing a loop does must be a conditional
+    const auto branch = as<IRIfElse>(nextBlock->getTerminator());
+    if(!branch)
+        return false;
+
+    if(!isSmallBlock(nextBlock))
+        return false;
+
+    const auto t = branch->getTrueBlock();
+    const auto f = branch->getFalseBlock();
+    const auto a = branch->getAfterBlock();
+
+    //
+    // In principle we could perform this simplification in the cfg simplifier,
+    // however it relies on slightly more context than is simple to insert
+    // there, namely that the removed trivial branching block is branching to a
+    // loop break block.
+    //
+
+    // Do we break on the 'true' side?
+    if(isSameBlockOrTrivialBranch(breakBlock, t) && f == a)
+    {
+        if(t != breakBlock)
+        {
+            branch->trueBlock.set(breakBlock);
+            t->removeAndDeallocate();
+        }
+        return true;
```

Comment:
Similarly to below, removing blocks like this can be subtle. Perhaps this could move to cfg simplification.

It changes

```
loop:
if cond goto d;
d: goto loopBreak;
loopBreak:
```
to
```
loop:
if cond goto loopBreak;
loopBreak:
```

When `d` is only used once with no arguments.

It's an enhancement in order to handle things like 

```
do{
	if(foo) break;
} while(x);
```

Reply:
Yeah, it might be beneficial to move this to cfg simplification. However we need to be careful not to remove critical edges or skipping merge points.